File: leds_unity/Assets/boubles/Shoots.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

namespace Boubles
{
    public class Shoots
    {
        int numLeds;
        public BoublesGame game;
        public List<ExplotionParticle> explotions;
        public List<Bullet> bullets;
        float delayToAdd = 3;

        public void Init(BoublesGame game, int numLeds)
        {
            this.game = game;
            explotions = new List<ExplotionParticle>();
            bullets = new List<Bullet>();
            this.numLeds = numLeds;
        }
        float timer = 0;

        public void Restart()
        {
            explotions = new List<ExplotionParticle>();
            bullets = new List<Bullet>();
        }

        public void OnUpdate(int centerLedID, int data1Count, int data2Count, float deltaTime)
        {
            foreach (ExplotionParticle e in explotions)
            {
                if (e.isOn)
                {
                    e.OnUpdate(deltaTime);
                    if (e.isOn)
                    {
                        Color c = e.color;
                        c.a = e.alpha;
                        game.ledsData[e.ledId] = c;
                    }
                }
            }

            int max = data1Count + centerLedID;
            int max2 = centerLedID - data2Count;
            foreach (Bullet e in bullets)
            {
                if (e.isOn)
                {
                    e.OnUpdate(deltaTime);
                    if(e.isOn)
                    {
                        foreach (int l in e.leds)
                        {
                            if ((e.characterID == 1 && l <= max) ||
                                (e.characterID == 2 && l >= max2)
                                )
                            {
                                game.CollideWith(e.color, e.characterID);
                                e.Collide();
                            }
                            else
                            {
                                game.ledsData[l] = game.colors[e.color];
                            }
                        }
                    }
                }
            }
        }       
        public void AddBullet(int characterID, int ledID, int color)
        {
            foreach (Bullet b in bullets)
            {
                if (!b.isOn)
                {
                    b.Restart(characterID, ledID, color);
                    return;
                }
            }
            Bullet e = new Bullet();
            e.Init(this, characterID, numLeds, ledID, color);
            bullets.Add(e);
        }
        public void AddExplotion(int from, int to, int characterID, int color)
        {
            for (int a = from; a < to; a++)
                AddExplotionParticle(a, characterID, color);
        }
        public void AddExplotionParticle(int ledID, int characterID, int color)
        {
            int dir = 1;
            if (characterID == 2)  dir = -1;
            Color c = game.colors[color];
            foreach (ExplotionParticle ex in explotions)
            {
                if (!ex.isOn)
                {
                    ex.Restart(ledID, dir, c);
                    return;
                }
            }
            ExplotionParticle e = new ExplotionParticle();
           
            e.Init(numLeds, ledID, dir, c);
            explotions.Add(e);
        } 
    }
}

File: micropython/bubbles/shoots.py
import explotion, bullet

class Shoots:
    def Init(self, game, numLeds):
        self.game = game
        self.explotions = []
        self.bullets = []
        self.numLeds = numLeds
        self.delayToAdd = 3
        self.timer = 0

    def Restart(self):
        self.explotions = []
        self.bullets = []

    def OnUpdate(self, centerLedID, data1Count, data2Count, deltaTime):
        for e in self.explotions:
            if e.isOn:
                e.OnUpdate(deltaTime)
                if e.isOn:
                    c = e.color
                    self.game.SetLed(e.ledId, c, e.alpha)
#                     self.game.ledsData[e.ledId] = c+e.alpha

        max_data1 = data1Count + centerLedID
        max_data2 = centerLedID - data2Count
        for e in self.bullets:
            color = e.color
            if e.isOn:
                e.OnUpdate(deltaTime)
                if e.isOn:
                    for l in e.leds:
                        if (e.characterID == 1 and l <= max_data1) or (e.characterID == 2 and l >= max_data2):
                            self.game.CollideWith(color, e.characterID)
                            e.Collide()
                            return
                        else:
                            self.game.SetLed(l, color)
#                             self.game.ledsData[l] = self.game.colors[color]

    def AddBullet(self, characterID, ledID, color):
        for b in self.bullets:
            if not b.isOn:
                b.Restart(characterID, ledID, color)
                return
        e = bullet.Bullet()
        e.Init(self, characterID, self.numLeds, ledID, color)
        self.bullets.append(e)

    def AddExplotion(self, from_range, to_range, characterID, color):        
        for a in range(from_range, to_range):
            self.AddExplotionParticle(a, characterID, color)

    def AddExplotionParticle(self, ledID, characterID, color):
        dir = 1 if characterID == 1 else -1
        c = self.game.colors[color]
        for ex in self.explotions:
            if not ex.isOn:
                ex.Restart(ledID, dir, c)
                return
        e = explotion.ExplotionParticle()
        e.Init(self.numLeds, ledID, dir, c)
        self.explotions.append(e)
